### DISCUSSION TITLE
Import Cycles

### DISCUSSION THREAD

**Discussion by lovetox** [microsoft/pyright#2926] [Q&A] [answered]
Hi,

I have a problem with import cycles, i tried different things, but pyright does always complain

I uploaded a test repo with a simple example

https://github.com/lovetox/testproject

Result:
```
Found 4 source files
/home/lovetox/projects/testproject/testproject/first_module.py
  /home/lovetox/projects/testproject/testproject/first_module.py:error: Cycle detected in import chain
    /home/lovetox/projects/testproject/testproject/first_module.py
    /home/lovetox/projects/testproject/testproject/second_module.py
1 error, 0 warnings, 0 informations 
```

i think pyright does not really check for **runtime** import cycles. because there are none. It seems to discover import cycles while type checking. But are those an issue?

As i see it its impossible to resolve these import cycles that only exist because of type hints and which TYPE_CHECKING should solve.

This is on pyright 1.1.214

**Comment by erictraut** [comment, marked as answer]
Runtime cycle detection is not generally possible to detect statically. It depends on temporal ordering of imports and symbol accesses.

Pyright's cycle detection was designed to detect architectural layering violations. In well-designed programs, modules should generally be strictly layered. If module A imports from module B, there should be no use of types defined in module A within module B. If there are, the code should be refactored to move the common types into the "lower" module or into some common module that is "lower" than both A and B.

This form of cycle detection may report some cases that do not result in runtime cycle violations.
